### ISSUE TITLE
banner does not appear in the chunk

### ISSUE THREAD

**Issue by gajus** [rollup/rollup#4850]
### Rollup Version

latest

### Operating System (or Browser)

N/A

### Node Version (if applicable)

N/A

### Link To Reproduction

N/A

### Expected Behaviour

Banner to appear in the chunk.

### Actual Behaviour

Does not appear.

```ts
banner: (chunk) => {
  const banner =
    '// moduleIds: \n' +
    chunk.moduleIds
      .map((moduleId) => {
        return (
          '// ' +
          moduleId
            .replaceAll(
              // eslint-disable-next-line no-control-regex
              /[\u0000-\u001F\u007F-\u009F]/g,
              '',
            )
            .trim()
            .replace(path.resolve(__dirname, '../..'), '@')
        );
      })
      .join('\n');

  // eslint-disable-next-line no-console
  console.log({ banner, imports: chunk.imports, name: chunk.name });

  return banner;
},
```

This is the code.

If I replace `banner` with some simple string like `// Hello, World!`, then it appears. Otherwise, it does not insert anything. Console logs confirm that it should.

**Comment by lukastaegert** [comment]
It seems to work for me. I tried to put it into the config file of `rollup` itself, and it added the appropriate banners. Note that I needed to change the line with `__dirname` to `.replace(fileURLToPath(new URL('../..', import.meta.url)), '@')` because there is not `__dirname` in ESM config files. For some entry points, the banner was only `// moduleIds:` because they were only facades, but that is expected.

**Comment by gajus** [comment]
@lukastaegert What happens if you try in the project replica I've shared with you?

**Comment by lukastaegert** [comment]
Yes, the comments do indeed vanish. I did some debugging and it is not Rollup that is failing here. The plugin `vite:esbuild-transpile` is removing the comments in its `renderChunk` hook. Not sure there is anything that can be done about as it may be that the issue is within esbuild. Still, you could post the issue to Vite. From Rollup side, there is definitely nothing that can be done about it.

**Comment by gajus** [comment]
Got it

**Comment by gajus** [comment]
It looks like ESBuild does not provide an option to keep comments, except legal comments.

https://github.com/evanw/esbuild/issues/221#issuecomment-653804184

However, using the above knowledge, a hack to keep comments is something like this:

```
banner: (chunk) => {
  const banner =
    '// @preserve moduleIds: \n' +
    chunk.moduleIds
      .map((moduleId) => {
        return (
          '// @preserve ' +
          moduleId
            .replaceAll(
              // eslint-disable-next-line no-control-regex
              /[\u0000-\u001F\u007F-\u009F]/g,
              '',
            )
            .trim()
            .replace(path.resolve(__dirname, '../..'), '@')
        );
      })
      .join('\n');

  // eslint-disable-next-line no-console
  console.log({ banner, imports: chunk.imports, name: chunk.name });

  return banner;
},
```

This was good enough for my debugging use case.
